src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const array = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(stateCopy, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (key in stateCopy) {
+            delete stateCopy[key];
```

Comment:
Using `delete` mutates `stateCopy`. Consider creating a new object without the keys to remove, e.g., `stateCopy = Object.fromEntries(Object.entries(stateCopy).filter(([key]) => !action.keysToRemove.includes(key)));`.